### ISSUE TITLE
firebase serve error: Cannot set headers after they are sent to the client

### ISSUE THREAD

**Issue by nheldman** [firebase/firebase-tools#2262]
### [REQUIRED] Environment info

**firebase-tools: 8.3.0**

**Platform: macOS**

### [REQUIRED] Test case

Please see the steps to reproduce below for a complete test case.

### [REQUIRED] Steps to reproduce

1. Previously was running `firebase-tools` version `8.2.0`
2. Ran `sudo npm i -g firebase-tools` to upgrade to `8.3.0`
3. Ran `firebase --version` to confirm upgrade to `8.3.0` was successful (it was)
4. Ran `firebase serve` in my local project, which I had run `firebase init` on weeks ago
5. Navigated to `localhost:5000/__/firebase/init.js` in the browser (file did not exist)
6. Local console displayed the following error:

```
Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (_http_outgoing.js:485:11)
    at /usr/local/lib/node_modules/firebase-tools/lib/hosting/initMiddleware.js:27:17
    at middleware (/usr/local/lib/node_modules/firebase-tools/lib/serve/hosting.js:52:17)
    at _run (/usr/local/lib/node_modules/firebase-tools/node_modules/superstatic/lib/activator.js:58:16)
    at /usr/local/lib/node_modules/firebase-tools/node_modules/superstatic/lib/responder.js:45:7
    at tryCatch (/usr/local/lib/node_modules/firebase-tools/node_modules/rsvp/dist/rsvp.js:525:12)
    at invokeCallback (/usr/local/lib/node_modules/firebase-tools/node_modules/rsvp/dist/rsvp.js:538:13)
    at publish (/usr/local/lib/node_modules/firebase-tools/node_modules/rsvp/dist/rsvp.js:508:7)
    at flush (/usr/local/lib/node_modules/firebase-tools/node_modules/rsvp/dist/rsvp.js:2415:5)
    at processTicksAndRejections (internal/process/task_queues.js:76:11)
```

7. Ran `sudo npm i -g firebase-tools@8.2.0` to downgrade
8. Ran `Ran `firebase --version` to confirm downgrade to `8.2.0` was successful (it was)
9. Ran `firebase serve` in my local project
10. Navigated to `localhost:5000/__/firebase/init.js` in the browser (file existed and loaded successfully)

### [REQUIRED] Expected behavior

When running `firebase-tools` version `8.3.0`, Javascript SDK files and init.js are served and load correctly.

### [REQUIRED] Actual behavior

When running `firebase-tools` version `8.3.0`, files are not served, and the error above is displayed

```
[2020-05-21T06:38:31.218Z] > command requires scopes: ["email","openid","https://www.googleapis.com/auth/cloudplatformprojects.readonly","https://www.googleapis.com/auth/firebase","https://www.googleapis.com/auth/cloud-platform"]
[2020-05-21T06:38:31.219Z] > authorizing via signed-in user
[2020-05-21T06:38:31.219Z] [iam] checking project fake-project-id for permissions ["firebase.projects.get"]
[2020-05-21T06:38:31.221Z] >>> HTTP REQUEST POST https://cloudresourcemanager.googleapis.com/v1/projects/fake-project-id:testIamPermissions  
 {"permissions":["firebase.projects.get"]}
[2020-05-21T06:38:31.587Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Thu, 21 May 2020 06:38:31 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","server-timing":"gfet4t7; dur=211","alt-svc":"h3-27=\":443\"; ma=2592000,h3-25=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q049=\":443\"; ma=2592000,h3-Q048=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
[2020-05-21T06:38:31.588Z] >>> HTTP REQUEST GET https://firebase.googleapis.com/v1beta1/projects/fake-project-id  
 
[2020-05-21T06:38:31.763Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Thu, 21 May 2020 06:38:31 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-27=\":443\"; ma=2592000,h3-25=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q049=\":443\"; ma=2592000,h3-Q048=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}

=== Serving from '/Users/noah.heldman/src/Client/fake-project-name'...

[2020-05-21T06:38:31.767Z] >>> HTTP REQUEST GET https://firebase.googleapis.com/v1beta1/projects/fake-project-id/webApps/-/config  
 
⚠  Your requested "node" version "8" doesn't match your global version "12" {"metadata":{"emulator":{"name":"functions"},"message":"Your requested \"node\" version \"8\" doesn't match your global version \"12\""}}
[2020-05-21T06:38:32.198Z] <<< HTTP RESPONSE 200 {"content-type":"application/json; charset=UTF-8","vary":"X-Origin, Referer, Origin,Accept-Encoding","date":"Thu, 21 May 2020 06:38:32 GMT","server":"ESF","cache-control":"private","x-xss-protection":"0","x-frame-options":"SAMEORIGIN","x-content-type-options":"nosniff","alt-svc":"h3-27=\":443\"; ma=2592000,h3-25=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q049=\":443\"; ma=2592000,h3-Q048=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"","accept-ranges":"none","transfer-encoding":"chunked"}
i  functions: Watching "/Users/noah.heldman/src/Client/fake-project-name/functions" for Cloud Functions... {"metadata":{"emulator":{"name":"functions"},"message":"Watching \"/Users/noah.heldman/src/Client/fake-project-name/functions\" for Cloud Functions..."}}
[2020-05-21T06:38:32.225Z] [worker-pool] addWorker(~diagnostic~) {"metadata":{"emulator":{"name":"functions"},"message":"[worker-pool] addWorker(~diagnostic~)"}}
[2020-05-21T06:38:32.225Z] [worker-pool] Adding worker with key ~diagnostic~, total=1 {"metadata":{"emulator":{"name":"functions"},"message":"[worker-pool] Adding worker with key ~diagnostic~, total=1"}}
[2020-05-21T06:38:32.226Z] [worker-pool] submitWork(triggerId=) {"metadata":{"emulator":{"name":"functions"},"message":"[worker-pool] submitWork(triggerId=)"}}
[2020-05-21T06:38:32.226Z] [worker-~diagnostic~-9a655d70-ab4a-4402-831f-b78d51a59f0f]: Assigning socketPath: /var/folders/s5/h8wdsyqd05584fz6d8xvh5f80000gn/T/fire_emu_18575.sock {"metadata":{"emulator":{"name":"functions"},"message":"[worker-~diagnostic~-9a655d70-ab4a-4402-831f-b78d51a59f0f]: Assigning socketPath: /var/folders/s5/h8wdsyqd05584fz6d8xvh5f80000gn/T/fire_emu_18575.sock"}}
[2020-05-21T06:38:32.226Z] [worker-~diagnostic~-9a655d70-ab4a-4402-831f-b78d51a59f0f]: BUSY {"metadata":{"emulator":{"name":"functions"},"message":"[worker-~diagnostic~-9a655d70-ab4a-4402-831f-b78d51a59f0f]: BUSY"}}
i  hosting: Serving hosting files from: build {"metadata":{"emulator":{"name":"hosting"},"message":"Serving hosting files from: \u001b[1mbuild\u001b[22m"}}
✔  hosting: Local server: http://localhost:5000 {"metadata":{"emulator":{"name":"hosting"},"message":"Local server: \u001b[4m\u001b[1mhttp://localhost:5000\u001b[22m\u001b[24m"}}
[2020-05-21T06:38:32.431Z] [runtime-status] [18575] Functions runtime initialized. {"cwd":"/Users/noah.heldman/src/Client/fake-project-name/functions","node_version":"12.14.1"} {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] Functions runtime initialized. {\"cwd\":\"/Users/noah.heldman/src/Client/fake-project-name/functions\",\"node_version\":\"12.14.1\"}"}}
[2020-05-21T06:38:32.432Z] [runtime-status] [18575] Disabled runtime features: undefined {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] Disabled runtime features: undefined"}}
[2020-05-21T06:38:32.434Z] [runtime-status] [18575] Resolved module firebase-admin {"declared":true,"installed":true,"version":"8.9.2","resolution":"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-admin/lib/index.js"} {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] Resolved module firebase-admin {\"declared\":true,\"installed\":true,\"version\":\"8.9.2\",\"resolution\":\"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-admin/lib/index.js\"}"}}
[2020-05-21T06:38:32.435Z] [runtime-status] [18575] Resolved module firebase-functions {"declared":true,"installed":true,"version":"3.6.1","resolution":"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-functions/lib/index.js"} {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] Resolved module firebase-functions {\"declared\":true,\"installed\":true,\"version\":\"3.6.1\",\"resolution\":\"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-functions/lib/index.js\"}"}}
[2020-05-21T06:38:32.436Z] [runtime-status] [18575] Outgoing network have been stubbed. [{"name":"http","status":"mocked"},{"name":"http","status":"mocked"},{"name":"https","status":"mocked"},{"name":"https","status":"mocked"},{"name":"net","status":"mocked"}] {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] Outgoing network have been stubbed. [{\"name\":\"http\",\"status\":\"mocked\"},{\"name\":\"http\",\"status\":\"mocked\"},{\"name\":\"https\",\"status\":\"mocked\"},{\"name\":\"https\",\"status\":\"mocked\"},{\"name\":\"net\",\"status\":\"mocked\"}]"}}
[2020-05-21T06:38:32.463Z] [runtime-status] [18575] Checked functions.config() {"config":{}} {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] Checked functions.config() {\"config\":{}}"}}
[2020-05-21T06:38:32.464Z] [runtime-status] [18575] Resolved module firebase-functions {"declared":true,"installed":true,"version":"3.6.1","resolution":"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-functions/lib/index.js"} {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] Resolved module firebase-functions {\"declared\":true,\"installed\":true,\"version\":\"3.6.1\",\"resolution\":\"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-functions/lib/index.js\"}"}}
[2020-05-21T06:38:32.465Z] [runtime-status] [18575] Resolved module firebase-admin {"declared":true,"installed":true,"version":"8.9.2","resolution":"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-admin/lib/index.js"} {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] Resolved module firebase-admin {\"declared\":true,\"installed\":true,\"version\":\"8.9.2\",\"resolution\":\"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-admin/lib/index.js\"}"}}
[2020-05-21T06:38:32.466Z] [runtime-status] [18575] Resolved module firebase-functions {"declared":true,"installed":true,"version":"3.6.1","resolution":"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-functions/lib/index.js"} {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] Resolved module firebase-functions {\"declared\":true,\"installed\":true,\"version\":\"3.6.1\",\"resolution\":\"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-functions/lib/index.js\"}"}}
[2020-05-21T06:38:32.466Z] [runtime-status] [18575] firebase-admin has been stubbed. {"adminResolution":{"declared":true,"installed":true,"version":"8.9.2","resolution":"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-admin/lib/index.js"}} {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] firebase-admin has been stubbed. {\"adminResolution\":{\"declared\":true,\"installed\":true,\"version\":\"8.9.2\",\"resolution\":\"/Users/noah.heldman/src/Client/fake-project-name/functions/node_modules/firebase-admin/lib/index.js\"}}"}}
[2020-05-21T06:38:32.468Z] [runtime-status] [18575] initializeApp(DEFAULT) {"databaseURL":"https://fake-project-id.firebaseio.com","storageBucket":"fake-project-id.appspot.com","projectId":"fake-project-id"} {"metadata":{"emulator":{"name":"functions"},"message":"[runtime-status] [18575] initializeApp(DEFAULT) {\"databaseURL\":\"https://fake-project-id.firebaseio.com\",\"storageBucket\":\"fake-project-id.appspot.com\",\"projectId\":\"fake-project-id\"}"}}
⚠  functions: The Cloud Firestore emulator is not running, so calls to Firestore will affect production. {"metadata":{"emulator":{"name":"functions"},"message":"The Cloud Firestore emulator is not running, so calls to Firestore will affect production."}}
⚠  External network resource requested!
   - URL: "http://169.254.169.254/computeMetadata/v1/instance"
 - Be careful, this may be a production service. {"metadata":{"emulator":{"name":"functions"},"message":"External network resource requested!\n   - URL: \"http://169.254.169.254/computeMetadata/v1/instance\"\n - Be careful, this may be a production service."}}
⚠  External network resource requested!
   - URL: "http://metadata.google.internal./computeMetadata/v1/instance"
 - Be careful, this may be a production service. {"metadata":{"emulator":{"name":"functions"},"message":"External network resource requested!\n   - URL: \"http://metadata.google.internal./computeMetadata/v1/instance\"\n - Be careful, this may be a production service."}}
[2020-05-21T06:38:32.821Z] [worker-~diagnostic~-9a655d70-ab4a-4402-831f-b78d51a59f0f]: IDLE {"metadata":{"emulator":{"name":"functions"},"message":"[worker-~diagnostic~-9a655d70-ab4a-4402-831f-b78d51a59f0f]: IDLE"}}
```

**Comment by samtstern** [comment]
@nheldman thanks for filing this issue.  The code you references is not included in `8.3.0`, it is in `master` but will be included in the next release (either `8.3.1` or `8.4.0`).

@abeisgoat can you triple-check that this issue is not present in `master`?

**Comment by nheldman** [comment]
I confirmed with my teammates that this issue occurs for another user when running on MacOS, but it does NOT happen for one of my Windows users. Hope this helps.

**Comment by nheldman** [comment]
While testing this issue, I saw that `8.4.0` was released. The issue is resolved for me in `8.4.0`, thanks!
